Convert productionlists in the Gallina chapter up to the Vernacular section to prodns (#11314)
coq/coq: doc/sphinx/language/gallina-specification-language.rst

  integer.  Underscores embedded in the digits are ignored, for example
  ``1_000_000`` is the same as ``1000000``.

  .. insertprodn numeral digit

  .. prodn::

>      numeral ::= {+ @digit } {? . {+ @digit } } {? {| e | E } {? {| + | - } } {+ @digit } }

Author: I think you're mistaken.  The `-` may be considered an operator rather than part of the numeral.

cLexer.ml has:

```
  | Some ('0'..'9') ->
      let n = NumTok.parse s in
      let ep = Stream.count s in
      comment_stop bp;
      (NUMERAL n, set_loc_pos loc bp ep)
```

and NumTok.parse doesn't look for a leading `-`.

`integer`, which gets renamed to `int` is defined as:

```
  integer:
    [ [ i = NUMERAL      -> { my_int_of_string loc (check_int loc i) }
      | "-"; i = NUMERAL -> { - my_int_of_string loc (check_int loc i) } ] ]
  ;
```

FWIW, the error message in `my_int_of_string` doesn't mention that decimal numbers are not permitted, either:

```
let my_int_of_string loc s =
  try
    int_of_string s
  with Failure _ ->
    CErrors.user_err ~loc (Pp.str "This number is too large.")
```

>      numeral ::= {+ @digit } {? . {+ @digit } } {? {| e | E } {? {| + | - } } {+ @digit } }

Reviewer: OK, then do you think that the text should be changed? And can you open a PR to fix the `my_int_of_string` issue you noticed (or at least a new issue)?

>      numeral ::= {+ @digit } {? . {+ @digit } } {? {| e | E } {? {| + | - } } {+ @digit } }

Author: I take it back, the `check_int` covers that case.  So nothing to change.  But OCaml has this little gem :-(:

```
# 4611686018427387904 = -4611686018427387904;;
- : bool = true
```

Because 

```
# 4611686018427387904;;
- : int = -4611686018427387904
```

Qualified identifiers and simple identifiers
--------------------------------------------

.. insertprodn qualid field_ident

.. prodn::
   qualid ::= @ident {* @field_ident }
   field_ident ::= .@ident

*Qualified identifiers* (:token:`qualid`) denote *global constants*

> (definitions, lemmas, theorems, remarks or facts), *global variables*

Reviewer: This paragraph contains a reference to `field`, which was renamed `field_ident`.

Definition by cases: match
--------------------------

> 

Author: Got it.

I notice that syntax embedded in the text doesn't stand out so well, e.g. "return".  Maybe something to tweak, but probably in another PR.  Maybe make it bold?  It is bold in the prodn table.

![image](https://user-images.githubusercontent.com/1253341/71327946-47804500-24c4-11ea-8244-8ce4a9347f27.png)

> 

Reviewer: I would like this, yes.

> 

Author: Done.  All notations are now bold.

.. insertprodn term_match pattern0

.. prodn::
   term_match ::= match {+, @case_item } {? return @term100 } with {? %| } {*| @eqn } end
   case_item ::= @term100 {? as @name } {? in @pattern }
   eqn ::= {+| {+, @pattern } } => @term
   pattern ::= @pattern10 : @term
   | @pattern10
   pattern10 ::= @pattern1 as @name
   | {+ @pattern1 }

>    | @ @qualid {* @pattern1 }

Author: Because that paragraph (and maybe the code) is incorrect?  There are only 2 productions in `pattern` that include `qualid` (renamed from `Prim.reference`):

```
  pattern:
    | "10" LEFTA
      | "@"; r = Prim.reference; lp = LIST0 NEXT ->
        { CAst.make ~loc @@ CPatCstr (r, Some lp, []) } ]
    | "0"
      [ r = Prim.reference                -> { CAst.make ~loc @@ CPatAtom (Some r) }
```

What should replace `@qualid {* @ident}` in the paragraph?

>    | @ @qualid {* @pattern1 }

Reviewer: Actually, the paragraph is correct, as it describes the simplest form of pattern matching, e.g. `Definition foo n := match n with O => true | S n => false end.` However, I didn't realize but the grammar as presented already allows to build such a construction thanks to the `{+ @pattern1 }` clause (`pattern1` can just be refined into `qualid`). So there is nothing to fix here.

However, I notice in passing that `{+ @pattern1 }` and `@pattern1` are redundant. The second one could be removed in favor of the former. And for more symmetry with the form `@qualid {* @pattern1 }`, `{+ @pattern1 }` could be displayed as `@pattern1 {* @pattern1 }`.

>    | @ @qualid {* @pattern1 }

Author: > However, I notice in passing that {+ @pattern1 } and @pattern1 are redundant.

I noticed that, but in this case I thought maybe it's worth keeping the pattern -> pattern10 -> pattern1 -> pattern0 chain very explicit.  I will change it, though.

> And for more symmetry with the form `@ @qualid {* @pattern1 }`, `{+ @pattern1 }` could be displayed as `@pattern1 {* @pattern1 }`.

I assume you meant `@ @pattern1 {* @pattern1 }`, right?  Wouldn't that incorrectly include strings in the form `@ @pattern0 % @ident`?

>    | @ @qualid {* @pattern1 }

Reviewer: > > 
> 
> I assume you meant `@ @pattern1 {* @pattern1 }`, right? Wouldn't that incorrectly include strings in the form `@ @pattern0 % @ident`?

I don't understand. No leading `@` of course, otherwise that would not be the same grammar. I don't understand your question either.

>    | @ @qualid {* @pattern1 }

Author: Got it.  My bad.

This paragraph describes the basic form of pattern matching. See
Section :ref:`Mult-match` and Chapter :ref:`extendedpatternmatching` for the description
of the general form. The basic form of pattern matching is characterized
by a single :token:`case_item` expression, an :token:`eqn` restricted to a
single :token:`pattern` and :token:`pattern` restricted to the form
:n:`@qualid {* @ident}`.

The expression :n:`match @term {? return @term100 } with
@pattern`:math:`_1` :n:`=> @term`:math:`_1` :n:`| … | @pattern`:math:`_n`

> :n:`=> @term`:math:`_n` :n:`end` denotes a

Reviewer: You could go further and fix the bad quoting practice here.

Author: I think the subscripts are not needed.  They're a bit pedantic. Buh-bye

Also odd to use both n and i as indices.

Reviewer: Usually `i` is used throughout the manual to denote a variable varying between `0` or `1` and `n`.

This paragraph describes the basic form of pattern matching. See
Section :ref:`Mult-match` and Chapter :ref:`extendedpatternmatching` for the description
of the general form. The basic form of pattern matching is characterized
by a single :token:`case_item` expression, an :token:`eqn` restricted to a
single :token:`pattern` and :token:`pattern` restricted to the form
:n:`@qualid {* @ident}`.

The expression :n:`match @term {? return @term100 } with

> @pattern__1 => @term__1 | … | @pattern__n => @term__n end` denotes a

Reviewer: Here you could have gone further and change this to:

```suggestion
{+| @pattern__i => @term__i } end` denotes a
```

> *pattern matching* over the term :token:`term`:math:`_0` (expected to be

Reviewer: `term__0` doesn't exist anymore in the expression above so this should be:

```suggestion
*pattern matching* over the term :token:`term` (expected to be
```

> of an inductive type :math:`I`). The terms :token:`term`:math:`_1`\ …\

Reviewer: ```suggestion
of an inductive type :math:`I`). The terms :n:`@term__i`
```

> :token:`term`:math:`_n` are the *branches* of the pattern matching

Reviewer: ```suggestion
are the *branches* of the pattern matching
```

(goes together with the previous suggestion)

> expression. Each of :token:`pattern`:math:`_i` has a form :token:`qualid`

Reviewer: ```suggestion
expression. Each of :n:`@pattern__i` has a form :n:`@qualid @ident`
```

> :token:`ident` where :token:`qualid` must denote a constructor. There should be

Reviewer: ```suggestion
where :token:`qualid` must denote a constructor. There should be
```

(goes together with the previous suggestion)

exactly one branch for every constructor of :math:`I`.

> The :n:`return @term100` gives the type returned by the whole match

Reviewer: ```suggestion
The :n:`return @term100` clause gives the type returned by the whole match
```

expression. There are several cases. In the *non dependent* case, all

> branches have the same type, and the :n:`return @term100` specifies that type.

Reviewer: ```suggestion
branches have the same type, and :n:`return @term100` specifies that type.
```

Author: I think this is better as-is.  "the" makes the reference to the previous sentence explicit.

Notice the previous sentence begins with "**The** ```:n:`return @term100` ```clause ...".

In this case, :n:`return @term100` can usually be omitted as it can be
inferred from the type of the branches [1]_.

In the *dependent* case, there are three subcases. In the first subcase,

> the type in each branch may depend on the exact value being matched in

Reviewer: "as ident" clause below could lose the quotes and use a proper `:n:` environment.

branch and the whole pattern matching expression has a type determined
by the specific dependencies in the type of the term being matched. This
dependency of the return type in the annotations of the inductive type
is expressed with a clause in the form

> :n:`in` :math:`I` :n:`_ … _ @pattern … @pattern`, where

Reviewer: Let's stop mixing math-style and notation-style, that makes no sense. Furthermore, let's avoid ellipses when we can use notations for repetition:

```suggestion
:n:`in @qualid {+ _ } {+ @pattern }`, where
```

> -  :math:`I` is the inductive type of the term being matched;

Reviewer: ```suggestion
-  :token:`qualid` is the inductive type of the term being matched;
```

(goes with the previous suggestion)

> -  the :n:`_` match the parameters of the inductive type: the

Reviewer: ```suggestion
-  the holes :n:`_` match the parameters of the inductive type: the
```

Recursive and co-recursive functions: fix and cofix
---------------------------------------------------

.. insertprodn term_fix term1_extended

.. prodn::
   term_fix ::= let fix @fix_body in @term
   | let cofix @cofix_body in @term
   | fix @fix_decls
   | cofix @cofix_decls
   fix_decls ::= @fix_body

>    | @fix_body with {+with @fix_body } for @ident

Author: Got it.  But it seems odd that the `for ident` only applies if you have multiple `fix_body`s.

![image](https://user-images.githubusercontent.com/1253341/71328209-3afdeb80-24c8-11ea-9836-26a215b7727b.png)

>    | @fix_body with {+with @fix_body } for @ident

Reviewer: I can explain: the reason of the `for` clause is simply that when defining mutually recursive fixpoints, one has to choose which one will be exposed as the function within this term. Another solution (which I would find more natural) would be for this syntax to actually define a record holding all the fixpoints.

   This warning is an error by default. It is caused by using a
   command with some attribute it does not understand.

> .. comment [1]

Reviewer: The footnote [2] remains numbered 2, which is a bit weird. Can we fix this? I.e. make the only remaining footnote be numbered 1?

Author: Got it.

Seems a bit primitive that the author has to manage numbering the footnotes.

Reviewer: I don't know if there are better ways to do this.

Reviewer 2: See https://docutils.sourceforge.io/docs/user/rst/quickref.html#footnotes ; You can use [#] as the number for autonumbering
